x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java

```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+        MultiSearchResponse.Item[] responses = p.getResponses();
+        int nextResponse = 0;
+        for (Sequence sequence : batchToCheck) {
+            boolean leading = true;
+            boolean discarded = false;
+            Timestamp lastLeading = null;
+            Timestamp firstTrailing = null;
+            for (int i = 0; i < criteria.size(); i++) {
+                SequenceCriterion criterion = criteria.get(i);
+                if (criterion.missing()) {
+                    SearchResponse response = responses[nextResponse++].getResponse();
+                    if (discarded) {
+                        continue; // consume all the responses for this sequence even if it's already discarded
+                    }
+                    SearchHit[] hits = response.getHits().getHits();
+                    if (leading) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        lastLeading = lastLeading == null || lastLeading.delta(hitTimestamp) < 0 ? hitTimestamp : lastLeading;
+                    } else if (trailing(i)) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        firstTrailing = firstTrailing == null || firstTrailing.delta(hitTimestamp) > 0 ? hitTimestamp : firstTrailing;
+                    } else {
+                        if (hits.length > 0) {
+                            discarded = true;
+                        }
+                    }
+                } else {
+                    leading = false;
+                }
+            }
+            if (discarded == false) {
+                int lastStage = criteria.size() - 1;
+                if ((firstTrailing == null && lastLeading == null)
+                    || (lastLeading == null && matcher.isMissingEvent(0))
+                    || (firstTrailing == null && matcher.isMissingEvent(lastStage))
+                    || (matcher.isMissingEvent(0) && matcher.isMissingEvent(lastStage) && biggerThanMaxSpan(lastLeading, firstTrailing))
+                    || (matcher.isMissingEvent(0)
+                        && matcher.isMissingEvent(lastStage) == false
+                        && biggerThanMaxSpan(lastLeading, sequence.ordinal().timestamp()))
+                    || (matcher.isMissingEvent(0) == false
+                        && matcher.isMissingEvent(lastStage)
+                        && biggerThanMaxSpan(sequence.startOrdinal().timestamp(), firstTrailing))
+
+                ) {
+                    matcher.addToCompleted(sequence);
+                }
+            }
+        }
+        checkMissingEvents(next, listener);
```

Comment:
Why are we calling this here? This potentially calls again `doCheckMissingEvents` which will have the same sequences to check... but `sequencesToCheck` Set should probably be empty if there were less sequences in the set than MAX_SEQUENCES_TO_CHECK_FOR_MISSING, but if there were more than MAX_SEQUENCES_TO_CHECK_FOR_MISSING then `sequencesToCheck` still has sequences to check for missing and this is when these are checked. This means that MAX_SEQUENCES_TO_CHECK_FOR_MISSING is a way to "paginate" the search for missing events through the set of gathered sequences?

Reply:
Yes, that's the idea. Since sequences can be pretty big (up to 255 terms if I'm not mistaken), each term can be a missing event to check, and we will check multiple sequences with a single request, the multi-query could grow significantly.

Reply:
Imo, the name `MAX_SEQUENCES_TO_CHECK_FOR_MISSING` is not descriptive enough, probably that's why I also suggested https://github.com/elastic/elasticsearch/pull/92348#discussion_r1190856136. The current name suggests this is an absolute maximum setting (like `LogicalPlanBuilder.MAX_SAMPLE_QUERIES` for example), but in fact it's a page size kind of check: "this number of sequences can be checked at a time". Maybe you can find a more descriptive name for it? `MISSING_EVENTS_SEQUENCES_CHECK_BATCH_SIZE`? (I think `BATCH_SIZE` is a good choice, not too fond of the rest of the name though but it's not bad). Wdyt?

Reply:
I like the "BATCH_SIZE" 👍